# Spam Detector

In [59]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [60]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

*My prediction is that the Random Forest Classifer model will be a better fit for this data and have higher scores.Random Forest Regression is preferred over linear regression when predicting numerical values because it offers greater accuracy and prediction stability.

## Split the Data into Training and Testing Sets

In [61]:
# Create the labels set `y` and features DataFrame `X`
y= data['spam']

X = data.copy().drop(columns='spam')



In [62]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()


spam
0    2788
1    1813
Name: count, dtype: int64

In [63]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test= train_test_split(X,y)

X_train.head()


word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
3262            0.22               0.00           0.07           0.0   
3401            0.00               0.07           0.15           0.0   
1265            0.00               0.00           0.00           0.0   
3551            0.00               0.00           0.00           0.0   
2031            0.00               0.00           0.00           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
3262           0.07            0.07               0.0                0.14   
3401           0.12            0.02               0.0                0.02   
1265           0.00            0.00               0.0                0.00   
3551           0.00            0.30               0.0                0.00   
2031           0.00            0.00               0.0                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
3262              0.0            0.36  ...                  0.00        0.041   
3401              0.0            0.12  ...                  0.43        0.037   
1265              0.0            0.00  ...                  0.00        0.000   
3551              0.0            0.00  ...                  0.00        0.000   
2031              0.0            0.00  ...                  0.00        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
3262        0.031        0.000        0.031        0.000        0.000   
3401        0.225        0.007        0.015        0.041        0.003   
1265        0.177        0.000        0.443        0.088        0.000   
3551        0.000        0.000        0.087        0.000        0.000   
2031        0.000        0.000        0.000        0.000        0.000   

      capital_run_length_average  capital_run_length_longest  \
3262                       1.912                          22   
3401                       2.198                          25   
1265                       1.693                          16   
3551                       2.740                          64   
2031                       1.000                           1   

      capital_run_length_total  
3262                       568  
3401                      2458  
1265                        83  
3551                       148  
2031                         6  

[5 rows x 57 columns]

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [64]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler= StandardScaler()


In [65]:
# Fit the Standard Scaler with the training data
scaler= scaler.fit(X_train)

In [66]:
# Scale the training data
X_train_scaled= scaler.transform(X_train)

# Scale the testing data
X_test_scaled= scaler.transform(X_test)


## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [67]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression

lr_model= LogisticRegression(random_state=1)

lr_model.fit(X_train_scaled, y_train)


print(f'Training Data Score: {lr_model.score(X_train_scaled, y_train),}')
print(f'Testing Data Score: {lr_model.score(X_test_scaled,y_test),}')



Training Data Score: (0.9327536231884058,)
Testing Data Score: (0.9200695047784535,)


In [68]:
# Make and save testing predictions with the saved logistic regression model using the test data
predictions= lr_model.predict(X_test_scaled)
# Review the predictions
predictions

array([0, 0, 0, ..., 0, 1, 1])

In [69]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_score(y_test, predictions)

0.9200695047784535

## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [70]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier
rfc_model= RandomForestClassifier(random_state=1)
rfc_model

rfc_model.fit(X_train_scaled,y_train)

print(f'Training Data Score: {rfc_model.score(X_train_scaled,y_train)}')
print(f'Testing Data Score:{rfc_model.score(X_test_scaled,y_test)}')
      



Training Data Score: 0.9991304347826087
Testing Data Score:0.9582971329278888


In [71]:
# Make and save testing predictions with the saved logistic regression model using the test data
testing_predictions= rfc_model.predict(X_test_scaled)
# Review the predictions
testing_predictions

array([0, 0, 0, ..., 0, 1, 1])

In [72]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_score(y_test, testing_predictions)


0.9582971329278888

## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

*Replace the text in this markdown cell with your answers to these questions.*

#Both models performed very well, I would like to favor the linear regression model just because it seems to be the simpilist go to model however once I performed the accuracy score it was evident that the Random Forest Classifier performed better. Random Forest Classifer had a higher accuracy score by 1.9% 